File: baekjoon/graph/6497/power.py
import sys

class t_edge():
    def __init__(self, _from, to, weight):
        self._from = _from
        self.to = to
        self.weight = weight

def find(v):
    if (parent[v] == v):
        return (v)
    parent[v] = find(parent[v])
    return parent[v]

def _union(x, y):
    x_parent = find(x)
    y_parent = find(y)

    if x_parent < y_parent:
        parent[y_parent] = x_parent
    else:
        parent[x_parent] = y_parent

def min_weight(v_cnt):
    total_weight = 0

    for i in range(0, v_cnt):
        parent[i] = i

    v_cnt -= 1
    for cur_edge in edges:
        if find(cur_edge._from) != find(cur_edge.to):
            _union(cur_edge._from, cur_edge.to)
            v_cnt -= 1
            total_weight += cur_edge.weight
        if v_cnt == 0:
            break

    return (total_weight)

if __name__ == '__main__':
    while True:
        v_cnt, e_cnt = map(int, sys.stdin.readline().split())
        if v_cnt == 0 and e_cnt == 0:
            break
        sum = 0
        edges = []
        for _ in range(e_cnt):
            _from, to, weight = map(int, sys.stdin.readline().split())
            sum += weight
            edges.append(t_edge(_from, to, weight))
        parent = [0 for _ in range(v_cnt)]
        edges.sort(key = lambda x:x.weight)
        print(sum - min_weight(v_cnt))

File: baekjoon/graph/6497/power.cpp
#include <iostream>
#include <vector>
#include <algorithm>

struct t_edge
{
    int from;
    int to;
    int weight;

    inline bool operator<(const t_edge &other) const {
        return (this->weight < other.weight);
    }
};

typedef std::vector<t_edge> t_vec;

const int   MAX = 200000;

int   v_cnt;
int   parent[MAX];

int find(int v)
{
    if (parent[v] == v)
        return (v);
    return parent[v] = find(parent[v]);
}

void    _union(int x, int y)
{
    int x_parent = find(x);
    int y_parent = find(y);

    if (x_parent < y_parent)
        parent[y_parent] = x_parent;
    else
        parent[x_parent] = y_parent;
}

int min_weight(t_vec &vec)
{
    int total_weight = 0;

    for (int i = 0; i < v_cnt; ++i)
        parent[i] = i;

    --v_cnt;
    int vec_size = vec.size();
    for (int i = 0; i < vec_size && v_cnt; ++i)
    {
        const t_edge &cur_edge = vec[i];

        if (find(cur_edge.from) != find(cur_edge.to))
        {
            _union(cur_edge.from, cur_edge.to);
            --v_cnt;
            total_weight += cur_edge.weight;
        }
    }
    return (total_weight);
}

int main(void)
{
    std::cin.tie(0)->sync_with_stdio(0);

    while (true)
    {
        t_vec  vec;
        t_edge edge;
        int    e_cnt, sum = 0;

        std::cin >> v_cnt >> e_cnt;
        if (v_cnt == 0 && e_cnt == 0)
            break ;

        while (e_cnt--)
        {
            std::cin >> edge.from >> edge.to >> edge.weight;
            sum += edge.weight;
            vec.push_back(edge);
        }

        std::sort(vec.begin(), vec.end());
        std::cout << sum - min_weight(vec) << '\n';
    }
    return (0);
}
